castra/core.py

```diff
@@ -424,6 +451,27 @@ def _categorize(categories, df):
                             index=df.index)
 
 
+def partitionby_none(buf, new):
+    """Repartition to ensure partitions don't split duplicate indices"""
+    if new.empty:
+        return [], buf
+    else:
+        end = buf.index[-1]
+        if end == new.index[0]:
```

Comment:
Slightly out of order records do occur.  If it's cheap then we might want to consider extending this to support the `[1, 2, 3, 4], [3, 4, 5, 6]` case as well.  This might just be changing the `==` sign to a `<=` sign.